### PR TITLE
Add the clustering in map and show the list of request of the cluster clicked

### LINKED ISSUE #195: Modify the mapScreen to handle the case of having two request on the same place
For now, if you have two request on the exact same location, you cannot make the difference between them (you don't see there are two request).
This task will resolve this problem and you will be able to click on the one you want

### PR DESCRIPTION
**Pull request by Funoma**
## ✨ Summary
Implement manual clustering system for map markers with dynamic zoom-based grouping and custom badge markers.
**User Story**: #196
Addresses #195

---

### Core Implementation
* Add manual clustering algorithm for map markers based on geographic distance
* Implement zoom-dependent clustering radius (adjusts from 300km at world view to 20m at street view)
* Custom marker badges displaying cluster count with Google Maps styling
* Density-based clustering algorithm prioritizing high-density points
* Clicking on a cluster show the list of request present in there (like in requestList)

### Clustering Features
* **Dynamic Radius Calculation**:
  * Automatically adjusts clustering distance based on zoom level (world view: 300km, street view: 20m)
  * Smooth transitions between zoom levels with progressive radius adjustment
* **Smart Grouping**:
  * Density-based algorithm: processes high-density areas first
  * Each request appears in exactly one cluster
  * Markers at identical locations automatically grouped
  * Nearby requests sorted by distance within clusters
* **Custom Marker Badges**:
  * Blue circular badges with white borders (Google Maps style)
  * Individual blue markers for single requests

### UI/UX Features
* **Interactive Markers**:
  * Click individual marker: Shows request details in bottom sheet
  * Click cluster badge: Shows list of all requests in that cluster
  * Smooth camera animation when selecting markers
  * Automatic zoom to selected location

### Map Utilities
* `calculateDistance()`: Haversine formula for accurate geographic distance calculation
* `calculateClusterCenter()`: Computes average position of cluster members
* `clusterRequestsByDistance()`: Main clustering algorithm with density optimization
* `getClusterRadiusForZoom()`: Dynamic radius calculation based on zoom level
* `createMarkerWithNumber()`: Custom bitmap generation for cluster badges

---

### Test Infrastructure
* Added test for the UI of the list of request
* Added comprehensive test suite for clustering utilities (`MapClusteringTest.kt`)
* Added marker bitmap generation tests (`MapClusterImageTest.kt`)

### Test Coverage
* **92% line coverage** achieved for all clustering and marker utilities
* Comprehensive edge case coverage

---

## 📸 Video
[Screen_recording_20251117_131717.webm](https://github.com/user-attachments/assets/3d380b21-81f3-4ae6-9ea9-bd9027c85206)

## 📸 Screenshots
Capture d&#39;écran 2025-11-17 131847
Capture d&#39;écran 2025-11-17 131857
Capture d&#39;écran 2025-11-17 131958

---

### Algorithm Complexity
* Time complexity: O(n²) for clustering (acceptable for typical request volumes)
* Space complexity: O(n) for storing clusters

### Design Decisions
* **Manual clustering over library**: Full control over algorithm and UX
* **Zoom-based radius**: Natural user experience as map scale changes
* **Density-first approach**: Ensures major clusters form correctly before edge cases
* **Custom badges**: Consistent styling with Google Maps platform conventions

---

### Known Limitations
* Clustering radius adjustment formula could be fine-tuned based on user feedback
* Large clusters (1000+ requests) may need pagination in list view
* Map performance may degrade with 10,000+ simultaneous requests (recommend server-side clustering at that scale)

### AI Usage Disclosure
⚠️ **AI tools were used** to assist with writing code and writing this PR.

### Usage of the library Cluster of Googlemap
* **I didn't use this library and made my own cluster management because of two reasons**: 
* The first one is because it didn't manage to have a cluster for request having the same location. Even with some strategy for handle this case (like the offset), it didn't work.
* The second issue is that one of the clusters wasn’t very good: two requests that were actually quite far apart (when zooming in, it was clear they shouldn’t have been grouped together) ended up in the same cluster. I’d prefer to have control over how the clustering is done, especially by being able to adjust the distance metric.

Closes #195

### CODE REVIEW

File: app/src/main/java/com/android/sample/ui/map/Map.kt [hunk 1]
```diff
@@ -129,7 +144,27 @@ fun calculateZoomLevel(markerCount: Int): Float {
...
+    zoomLevel < ConstantMap.MAX_ZOOM_ONE -> ConstantMap.ZOOM_LEVEL_WORLD // 300km - mondial view
+    zoomLevel < ConstantMap.MAX_ZOOM_TWO -> ConstantMap.ZOOM_LEVEL_WL // 70km
+    zoomLevel < ConstantMap.MAX_ZOOM_THREE -> ConstantMap.ZOOM_LEVEL_LAND // 40km - land view
+    zoomLevel < ConstantMap.MAX_ZOOM_FOUR -> ConstantMap.ZOOM_LEVEL_REGION // 10km - region view
+    zoomLevel < ConstantMap.MAX_ZOOM_FIVE -> ConstantMap.ZOOM_LEVEL_CITY // 3km - city view
+    zoomLevel < ConstantMap.MAX_ZOOM_SIX -> ConstantMap.ZOOM_LEVEL_MID // 400m - view mid
+    zoomLevel < ConstantMap.MAX_ZOOM_SEVEN -> ConstantMap.ZOOM_LEVEL_STREET_BIG // 50m - view street
+    else -> ConstantMap.ZOOM_LEVEL_STREET_SMALL // 20m - vue street
+  } / (zoomLevel / ConstantMap.ZOOM_DIVIDE) // change with current zoom
+}
```
Comment (unusedusername01) [inline app/src/main/java/com/android/sample/ui/map/Map.kt]: > [!NOTE]
> I can feel the passion you put into this, great work and great documentation !

File: app/src/main/java/com/android/sample/ui/map/Map.kt [hunk 2]
```diff
@@ -815,6 +784,340 @@ fun Date.toDisplayStringWithoutHours(): String {
...
+ */
+fun createMarkerBitmap(count: Int): Bitmap {
+  val size = 100
+  val bitmap = createBitmap(size, size)
+  val canvas = Canvas(bitmap)
+
+  // Circle blue
+  val circlePaint =
+      Paint().apply {
+        color = "#4285F4".toColorInt() // Here I use directly color because I create a marker
```
Comment (unusedusername01) [inline app/src/main/java/com/android/sample/ui/map/Map.kt]: > [!NOTE]
> You indeed used a magic value for a STYLING COLOR ✖️ But you justified it somehow ✅ I won't come to your house tonight, but you better watch your back next time 👁️

File: app/src/main/java/com/android/sample/ui/map/ConstantMap.kt [hunk 3]
```diff
@@ -90,4 +96,45 @@ object ConstantMap {
...
+  const val NUMBER_SIZE_TWO = 32f
+  const val NUMBER_SIZE_THREE = 28f
+
+  // ======================
+  // Distance request
+  // ======================
+
+  const val EARTH_RADIUS = 6371000.0
+  const val TWO = 2
 }
```
[the remark this answers is not in the document]
Comment (Funoma) [reply app/src/main/java/com/android/sample/ui/map/ConstantMap.kt]: Thank you so much ! 😘

File: app/src/main/java/com/android/sample/ui/map/Map.kt [hunk 4]
```diff
@@ -178,20 +222,52 @@ fun MapScreen(viewModel: MapViewModel = viewModel(), navigationActions: Navigati
...
+                            cameraPositionState.animate(
+                                update =
+                                    CameraUpdateFactory.newLatLngZoom(
+                                        position,
+                                        zoomLevel + ConstantMap.ZOOM_LEVEL_TWO // Zoom 2 level
+                                        ),
+                                durationMs = ConstantMap.LONG_DURATION_ANIMATION)
+                            viewModel.updateCurrentListRequest(cluster)
+                          }
+                        }
```
[the remark this answers is not in the document]
Comment (Funoma) [reply app/src/main/java/com/android/sample/ui/map/Map.kt]: Just put a value which is equal to the condition (is a boolean)

File: app/src/main/java/com/android/sample/ui/map/Map.kt [hunk 5]
```diff
@@ -815,6 +784,340 @@ fun Date.toDisplayStringWithoutHours(): String {
...
+}
+
+/**
+ * Creates a bitmap for a marker with a number badge.
+ *
+ * @param count The number to display on the badge
+ * @return A Bitmap containing a blue circle with white border and the count number
+ */
+fun createMarkerBitmap(count: Int): Bitmap {
+  val size = 100
```
[the remark this answers is not in the document]
Comment (Funoma) [reply app/src/main/java/com/android/sample/ui/map/Map.kt]: Oops my bad 😅 
I really didn't want you to come to my place, which is why I didn't deliberately leave that magic value, and I wasn't disappointed that you weren't there tonight, but anyway... 😞 
Now it's fixed...

File: app/src/test/java/com/android/sample/map/MapClusterImageTest.kt
[the remark this answers is not in the document]
Comment (Funoma) [reply app/src/test/java/com/android/sample/map/MapClusterImageTest.kt]: Add constant, now it's should be good
